sound/soc/sof/ipc.c

```diff
@@ -444,8 +457,22 @@ static void ipc_period_elapsed(struct snd_sof_dev *sdev, u32 msg_id)
 	memcpy(&stream->posn, &posn, sizeof(posn));
 
 	/* only inform ALSA for period_wakeup mode */
-	if (!stream->substream->runtime->no_period_wakeup)
-		snd_pcm_period_elapsed(stream->substream);
+	if (!stream->substream->runtime->no_period_wakeup) {
+		if (stream->substream->runtime->status->state ==
+		    SNDRV_PCM_STATE_RUNNING) {
+			snd_pcm_period_elapsed(stream->substream);
+		} else {
+			/*
+			 * Other than running state might cause ipc back to
+			 * dsp. We need to defer processing because we might
+			 * otherwise put irq handler to sleep with ipc busy
+			 * bit enabled (waiting for dsp ack).
```

Comment:
that looks horribly Intel-specific and should probably be done somewhere else.

Reply:
yes agree totally, just want your comments, because I see deferred processing the only choice to solve this. It's probably also byt specific...